parsing/lexer.mll

```diff
@@ -420,33 +469,52 @@ rule token = parse
           (Reserved_sequence (".~", Some "is reserved for use in MetaOCaml")) }
   | "~" raw_ident_escape (lowercase identchar * as name) ':'
       { LABEL name }
-  | "~" (lowercase identchar * as name) ':'
+  | "~" raw_ident_escape (ident_ext as raw_name) ':'
+      { let name = ident_for_extended lexbuf raw_name in
+        if UString.is_capitalized name
+        then error lexbuf (Capitalized_label name);
+        LABEL name }
+  | "~" (identstart identchar * as name) ':'
       { check_label_name lexbuf name;
         LABEL name }
-  | "~" (lowercase_latin1 identchar_latin1 * as name) ':'
-      { warn_latin1 lexbuf;
+  | "~" (ident_ext as raw_name) ':'
+      { let name = ident_for_extended lexbuf raw_name in
+        check_label_name lexbuf name;
         LABEL name }
   | "?"
       { QUESTION }
   | "?" raw_ident_escape (lowercase identchar * as name) ':'
       { OPTLABEL name }
+  | "?" raw_ident_escape (ident_ext as raw_name) ':'
+      { let name = ident_for_extended lexbuf raw_name in
+        if UString.is_capitalized name
```

Comment:
The previous rule matches names that always pass this check. Why duplicate the rule ?

Going further, perhaps we could avoid using `identstart` and `identchar` entirely, ensuring that we support "extended" identifiers everywhere.

Reply:
This is a performance vs maintainability tradeoff. With the fast ascii path separate from the extended unicode path, we keep the same performance as the current lexer on micro-benchmarks. Merging the two paths in one results into a 2x slowdown on the same benchmark (see https://github.com/Octachron/ocaml-lexer-microbenchmarks).

Reply:
I see that the difference goes down to 1.30x with an extremely large and more representative example (all .ml files of this repo concatenated). Is lexing time that important ?

Reply:
Thinking about it maybe a better compromise would be to keep the fast ascii rules for identifiers and labels, but merge the raw_identifier case with the unicode one? This reduces the number of rules to:
- 3 for identifiers: lowercase_ascii, uppercase_ascii, complex
- 2 for labels and optional labels: lowercase_ascii, complex
with (most probably) no bearing on the performances.

Reply:
I think reducing the number of rules can only be good for maintenance. Do uppercase ASCII identifiers need a fast path too ?

Reply:
My opinion is that the ascii rule are also quite simpler to read:
```ocaml
  | lowercase identchar * as name
      { try Hashtbl.find keyword_table name
        with Not_found -> LIDENT name }
  | uppercase identchar * as name
      { UIDENT name } (* No capitalized keywords *)
```
compared to
```ocaml
  | (raw_ident_escape? as escape) (ident_ext as raw_name)
      { let name = ident_for_extended lexbuf raw_name in
        if UIdent.is_capitalized name then begin
            if escape="" then UIDENT name
            else error lexbuf (Capitalized_raw_identifier name)
        end else
          LIDENT name
      } (* No non-ascii keywords *)
 ```
Consequently, I belive that keeping both rules for the ascii case is fine (or possibly very slightly better?) in term of maintenance.

Reply:
I think in this example the rule is made complicated by the `Capitalized_raw_identifier` check. It could be defined elsewhere, like it's done for other checks.
The `is_capitalized` if seems as readable as two separate rules, though it's nice that the keyword lookup is done in a separate rule. Also, the more complex rule is not made even more complex by dropping the simpler rule.